File: covilo-core/src/main/java/me/kvdpxne/covilo/application/dto/UserDto.java
package me.kvdpxne.covilo.application.dto;

import java.time.LocalDateTime;
import me.kvdpxne.covilo.domain.model.City;
import me.kvdpxne.covilo.domain.model.Gender;

import java.time.LocalDate;
import java.util.UUID;

public record UserDto(
  UUID identifier,
  String email,
  String firstName,
  String lastName,
  Gender gender,
  LocalDate birthDate,
  CityDto livingPlace,
  LocalDateTime createdDate,
  LocalDateTime lastModifiedDate
) {
}

File: covilo-core/src/main/kotlin/me/kvdpxne/covilo/api/rest/dto/UserDto.kt
package me.kvdpxne.covilo.api.rest.dto

import me.kvdpxne.covilo.domain.model.User
import me.kvdpxne.covilo.domain.model.Users

/**
 * Shortcut to the [UserDto] collection.
 */
typealias UsersDto = Collection<UserDto>

/**
 * TODO doc
 */
data class UserDto(
  val identifier: String,
  val email: String,
  val firstName: String,
  val lastName: String,
  val brithDate: String,
  val createdDate: String,
  val lastModifiedDate: String
)

/**
 * Transforms the [User] entity to [UserDto].
 */
fun User.toDto(): UserDto {
  return UserDto(
    this.identifier.toString(),
    this.email,
    this.firstName.toString(),
    this.lastName.toString(),
    this.birthDate.toString(),
    this.createdDate.toString(),
    this.lastModifiedDate.toString()
  )
}

/**
 * Transforms all [User] entities that are in the collection
 * and are not null to [UserDto].
 */
fun Users.toDto(): UsersDto {
  return mapNotNull {
    it.toDto()
  }
}